peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+the same function definition:
+
+::
+
+   def as_instance[T](typx: TypeExpr[T]) -> T | None:
+       return typx() if isinstance(typx, type) else None
+
+A ``TypeExpr[]`` value corresponding to a
+:ref:`non-universal TypeExpr[] <non_universal_typeexpr>` may be
+treated as equivalent to ``Any`` when constraining a type variable:
+
+::
+
+   value = as_instance(TypeIs[str])
+   assert_type(value, Any)  # acceptable
+
+
+Interactions with type[]
+------------------------
+
+Both ``TypeExpr[]`` and ``type[]`` can be used to constrain the same type
+variable within the same function definition:
+
+::
+
+   def as_type[T](typx: TypeExpr[T]) -> type[T] | None:
+       return typx if isinstance(typx, type) else None
+
+
+Interactions with TypeIs[] and TypeGuard[]
+------------------------------------------
+
+A type variable constrained by ``TypeExpr[]`` can also be used by a ``TypeIs[]``
+within the same function definition:
+
+::
+
+   def isassignable[T](value: object, typx: TypeExpr[T]) -> TypeIs[T]: ...
+
+   count: int | str = ...
+   if isassignable(count, int):
+       assert_type(count, int)
+   else:
+       assert_type(count, str)
+
+or by a ``TypeGuard[]`` within the same function definition:
+
+::
+
+   def isdefault[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]:
+       return (value == typx()) if isinstance(typx, type) else False
+
+   value: int | str = ''
+   if isdefault(value, int):
+       assert_type(value, int)
+       assert 0 == value
+   elif isdefault(value, str):
+       assert_type(value, str)
+       assert '' == value
+   else:
+       assert_type(value, int | str)
+
+
+.. _interactions_with_annotated:
+
+Interactions with Annotated[]
+-----------------------------
+
+``Annotated[]`` forms preserve their metadata at runtime:
+
+::
+
+   >>> ValueRange: TypeAlias = slice
+   >>> PositiveInt: TypeAlias = Annotated[int, ValueRange(1, float('inf'))]
+   >>> PositiveInt
+   typing.Annotated[int, slice(1, inf, None)]  # NOT: int
+
+However ``Annotated[]`` forms need not preserve their metadata at
+typechecking-time:
+
+::
+
+   count: PositiveInt = 1
+   assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+
+In particular when an ``Annotated[]`` argument is passed to a ``TypeExpr[]``
+parameter constraining a type variable that is also used by a ``TypeIs[]``
+or ``TypeGuard[]``, only the first type argument to ``Annotated[]`` must be
+maintained in the type inferred by a type checker. The second metadata
+argument to ``Annotated[]`` does need not be maintained:
+
+::
+
+   # Similar to isassignable(), but accepts Annotated[] forms describing constraints
+   def ismatch[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]: ...
+
+   count: int | str = -1
+   if ismatch(count, PositiveInt):
+       assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+   else:
+       assert_type(count, int | str)
+
+The above behavior of stripping the metadata argument to ``Annotated[]`` is
```

Comment:
But type checkers such as mypy don't preserve Annotated at all, even when used directly to annotate a variable (`x: PositiveInt`). I don't see why TypeExpr is special here or why this behavior needs to be specified at all.

Reply:
In early discussions around TypeExpr (then: TypeForm) there seemed to be some concern - from the runtime type checker folks I think - that `Annotated[...]` metadata might be stripped out early due to this PEP. But that's actually the status quo anyway.

Rereading the entire §"Interactions with Annotated[]" section again, I don't think it says anything useful. Improving how `Annotated[]` interacts with the rest of the type system - and in particular preserving metadata under certain circumstances - is outside the scope of this PEP. I think I'll delete the section entirely.